compiler/rustc_lint/src/dangling.rs

```diff
@@ -0,0 +1,252 @@
+    /// The `dangling_pointers_from_temporaries` lint detects getting a pointer to data
+    /// of a temporary that will immediately get dropped.
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// # #![allow(unused)]
+    /// # unsafe fn use_data(ptr: *const u8) { }
+    /// fn gather_and_use(bytes: impl Iterator<Item = u8>) {
+    ///     let x: *const u8 = bytes.collect::<Vec<u8>>().as_ptr();
+    ///     unsafe { use_data(x) }
+    /// }
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// Getting a pointer from a temporary value will not prolong its lifetime,
+    /// which means that the value can be dropped and the allocation freed
+    /// while the pointer still exists, making the pointer dangling.
+    /// This is not an error (as far as the type system is concerned)
+    /// but probably is not what the user intended either.
+    ///
+    /// If you need stronger guarantees, consider using references instead,
+    /// as they are statically verified by the borrow-checker to never dangle.
+    pub DANGLING_POINTERS_FROM_TEMPORARIES,
+    Warn,
+    "detects getting a pointer from a temporary"
+}
+
+type Buffered<'tcx> = FxHashMap<HirId, Decorator<'tcx>>;
+
+/// FIXME: false negatives (i.e. the lint is not emitted when it should be)
+/// 1. Method calls that are not checked for:
+///    - [`temporary_unsafe_cell.get()`][`core::cell::UnsafeCell::get()`]
+///    - [`temporary_sync_unsafe_cell.get()`][`core::cell::SyncUnsafeCell::get()`]
+/// 2. Ways to get a temporary that are not recognized:
+///    - `owning_temporary.field`
+///    - `owning_temporary[index]`
+/// 3. No checks for ref-to-ptr conversions:
+///    - `&raw [mut] temporary`
+///    - `&temporary as *(const|mut) _`
+///    - `ptr::from_ref(&temporary)` and friends
+#[derive(Default)]
+pub(crate) struct DanglingPointers<'tcx> {
+    buffered: Buffered<'tcx>,
+}
+
+impl_lint_pass!(DanglingPointers<'_> => [DANGLING_POINTERS_FROM_TEMPORARIES]);
+
+// This skips over const blocks, but they cannot use or return a dangling pointer anyways.
+//
+// We cannot emit lints in the Visitor, because then `#[allow]` & `#[deny]` won't be honored.
+// So we buffer them instead.
+impl<'tcx> LateLintPass<'tcx> for DanglingPointers<'tcx> {
+    fn check_fn(
+        &mut self,
+        cx: &LateContext<'tcx>,
+        _: FnKind<'tcx>,
+        _: &'tcx FnDecl<'tcx>,
+        body: &'tcx Body<'tcx>,
+        _: Span,
+        _: LocalDefId,
+    ) {
+        Collector { cx, buffer: &mut self.buffered, suppress: false }.visit_body(body)
+    }
+
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>) {
+        if let Some(decorator) = self.buffered.remove(&expr.hir_id) {
+            // FIXME: use `emit_lint` when `#[primary_span]` is added.
+            cx.emit_span_lint(
+                DANGLING_POINTERS_FROM_TEMPORARIES,
+                decorator.primary_span(),
+                decorator,
+            )
```

Comment:
I think using [`TyCtxt::emit_node_span_lint`](https://doc.rust-lang.org/stable/nightly-rustc/rustc_middle/ty/context/struct.TyCtxt.html#method.emit_node_span_lint), which directly takes a `HirId` should remove the need to buffer the decorator at all, as it will compute the lint level at the given `HirId`.

In the visitor, something like this should do it:
```rust
cx.tcx.emit_span_lint(
    DANGLING_POINTERS_FROM_TEMPORARIES,
    expr.hir_id,
    expr.span,
    decorator,
)
```
which I think would also remove the need for the `check_fn_post`.